### PR TITLE
Flexible sequencing training

### PR DESCRIPTION
**Pull request by UnicornJessi**
Replacing lazy sequence generator with flexible sequence generator.

dataset is still shuffled but every song is used at least once per epoch

Processes .npz files cointaining continuous_sequence data
enabling training with different sequence length without new processing

Added for model preset: 
stride: Controls overlap between extracted subsequences
sequence_length: Overrides the global SEQUENCE_LENGTH constant per preset
batch size: Number of samples per training batch

### CODE REVIEW

File: src/groove_panda/config.py [hunk 1]
```diff
@@ -10,11 +10,11 @@ class TokenizeMode(Enum):
 
 # Hyperparameters
```
Comment (julienpoeschl) [inline src/groove_panda/config.py:11]: Nimm mal komplettes `config.py` aus der Pull Request raus, außer, du hast neue settings hinzugefügt, die alle haben sollten. Sonst wird bei jeder PR das `config.py` immer neu überschrieben

File: src/groove_panda/models/train.py [hunk 2]
```diff
@@ -6,38 +6,41 @@
...
 from groove_panda.models import plot
-from groove_panda.models.lazy_sequence_generator import LazySequenceGenerator
 from groove_panda.models.models import BaseModel
 from groove_panda.models.tf_custom.callbacks import TerminalPrettyCallback
 
 logger = logging.getLogger(__name__)
 
 
-def train_model(model: BaseModel, file_paths: list):
+def train_model(model: BaseModel, train_generator):
```
Comment (julienpoeschl) [inline src/groove_panda/models/train.py:15]: Zuvor wurde hier von Jan ja eine tensorflow lastige lösung präsentiert, die den `lazy_sequence_generator` ersetzt hat. Jetzt benutzt du ja wieder ein ähnliches neues objekt: Was musstest du dafür ersetzen oder rauswerfen? Eventuell mal Jan fragen, damit er sich das auch mal angucken kann

Comment (UnicornJessi) [reply src/groove_panda/models/train.py:15]: huh was war jans lösung xD

File: src/groove_panda/config.py (hunk 1 above)
Comment (julienpoeschl) [reply src/groove_panda/config.py]: top

File: src/groove_panda/models/train.py (hunk 2 above)
Comment (julienpoeschl) [reply src/groove_panda/models/train.py:15]: Aktuell im code sollte, glaube ich, `train_model_eager` verwendet werden. `train_model` in `train.py` ist eine veraltete version, die behalten wurde, falls die andere nicht klappt

File: src/groove_panda/processing/process.py [hunk 3]
```diff
@@ -73,6 +73,46 @@ def numerize(sixtuples: list[Sixtuple], sixtuple_token_maps: SixtupleTokenMaps)
     return numeric_sixtuples
 
 
+def create_continuous_sequence(numeric_sixtuples: list[NumericSixtuple]) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (UnicornJessi) [reply src/groove_panda/processing/process.py:76]: huh? es ist handle_train -> controller: train -> model_training: train_model -> train: train_model

[the remark this answers is not in the document]
Comment (UnicornJessi) [reply src/groove_panda/processing/process.py:76]: ik ich müsste das aber anpassen/ nochmal genauer anschauen

[the remark this answers is not in the document]
Comment (julienpoeschl) [reply src/groove_panda/processing/process.py:76]: Ist das in Bearbeitung? Vllt liege ich auch falsch und sehe das nicht richtig, aber für mich sah es so aus, als ob wir sonst die letzten implementierungen von jan rauswerfen würden, was glaube ich nicht sein muss. Kann man deine neuen features mit dem jetzigen verbinden?

[the remark this answers is not in the document]
Comment (UnicornJessi) [reply src/groove_panda/processing/process.py:76]: habs auf eager geändert

File: src/groove_panda/models/model_training.py [hunk 4]
```diff
@@ -34,12 +34,21 @@ def train_model(model_id: str, processed_dataset_id: str, preset_name: str):
...
 
-    # Get input shape from first file
-    with np.load(file_paths[0]) as data:
-        input_shape = data["x"].shape[1:]
+    # Get preset config
+    preset = MODEL_PRESETS[preset_name]
+    sequence_length = preset["sequence_length"]
 
-    model = models.LSTMModel(model_id, input_shape)
+    model = models.LSTMModel(model_id, (sequence_length, 6))
```
Comment (julienpoeschl) [inline src/groove_panda/models/model_training.py:41]: Kann man die 6 hier (Anzahl features) von irgendwo holen, statt sie als magic number zu lassen?

File: src/groove_panda/models/model_training.py [hunk 5]
```diff
@@ -34,12 +34,21 @@ def train_model(model_id: str, processed_dataset_id: str, preset_name: str):
...
     model.build(vocab_sizes=vocab_sizes, preset_name=preset_name)
 
-    tr.train_model_eager(model, file_paths)
+    # Use flexible sequence generator instead of loading all data
+    train_generator = FlexibleSequenceGenerator(
+        file_paths=file_paths,
+        sequence_length=sequence_length,
+        stride=preset["stride"],
+        batch_size=preset.get("batch_size", 32),
```
Comment (julienpoeschl) [inline src/groove_panda/models/model_training.py:49]: Im `config.py` sollte ein eintrag für batch size existieren. Bitte verwende diesen, statt zu hardcoden

File: src/groove_panda/models/train.py [hunk 6]
```diff
@@ -61,19 +66,33 @@ def train_model_eager(model: BaseModel, file_paths: list):
...
-        logger.info("Start gathering processed songs...")
-        # Concatenates the sample seq["bar", "position", "pitch", "duration", "velocity", "tempo"])uences of all files
-        # into an array
-        full_array_x = np.concatenate([(np.load(path))["x"] for path in file_paths])
-        # Concatenates the sample targets of all files into an array
-        full_array_y = np.concatenate([(np.load(path))["y"] for path in file_paths])
-        # Ensures that the lengths of samples of targets match
+        logger.info("Start gathering all subsequences from continuous data...")
+
+        # Extract ALL possible subsequences using FlexibleSequenceGenerator logic
```
Comment (julienpoeschl) [inline src/groove_panda/models/train.py:71]: Hast du schon getestet, wie sich das von der Geschwindigkeit im Vergleich zu vorher verhält? Gleich, schneller, langsamer?

File: src/groove_panda/models/model_training.py (hunk 5 above)
Comment (UnicornJessi) [reply src/groove_panda/models/model_training.py:49]: ok habs geändert. also ohne die konstanten xD

File: src/groove_panda/models/train.py (hunk 6 above)
Comment (UnicornJessi) [reply src/groove_panda/models/train.py:71]: es ist ungefähr gleich schnell vllt 30 sek länger höchstens

File: src/groove_panda/processing/process.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (UnicornJessi) [reply src/groove_panda/processing/process.py:76]: ich hab tatsächlich glaube ich keine methoden gelöscht weil wir lazy sequence generator behalten haben und du gemeint hast wir lassen das erstmal drinne also dachte ich lasse ich die methoden die lazy sequence braucht auch mal so wie sie sind

[the remark this answers is not in the document]
Comment (julienpoeschl) [reply src/groove_panda/processing/process.py:76]: Passt. Wir werden demnächst alles Ungebrauchtes entfernen

File: src/groove_panda/models/train.py (hunk 6 above)
Comment (julienpoeschl) [reply src/groove_panda/models/train.py:71]: Pro Epoche oder gesamt? Für wie viele songs?

File: src/groove_panda/models/model_training.py (hunk 5 above)
Comment (julienpoeschl) [reply src/groove_panda/models/model_training.py]: Ich hatte was falsch gelesen und nicht gekocht, aber die Änderung ist glaube ich auch besser so.
